### PR TITLE
direct mapping: misc fixes

### PR DESCRIPTION
**Pull request by alessandrod**
This fixes the following issues with the direct mapping feature, it has no impact on !direct_mapping code:

- `bpf_loader: direct_mapping: enforce account info pointers to be immutable` ensures that all AccountInfo pointers are not mutated. CPI looks up memory regions based on an account's data address so it can't be changed, and other pointers are locked too since there's no (legitimate) reason to ever mutate them, they point to address space setup by the program runtime.

- `bpf_loader: cpi: access ref_to_len_in_vm through VmValue` disables "caching" of vm->host address translations. Vm address space under direct mapping has some dynamic properties (CoW, changing permissions), so memory accesses must always to go through the vm.

- `direct mapping: improve memory permission tracking across CPI calls` ensures that account data and realloc memory regions are correctly updated to match their respective account's permissions, and that the permissions are all synchronized in one go _before_ the individual accounts are synchronized. This is required to guarantee that while synchronizing accounts across CPI calls memory region permissions are in a consistent state.

 - `cpi: fix slow edge case zeroing extra account capacity after shrinking an account` fixes a slow edge case when resizing large accounts. 

- ` cpi: direct_mapping: error out if ref_to_len_in_vm points to account memory` forbids `AccountInfo::data` from being put in account memory. CPI updates that slice and we don't want CPI to write into account data.

- `bpf_loader: direct_mapping: map AccessViolation -> InstructionError` maps low level `AccessViolation` errors to higher level `InstructionError`s for better debuggability

Extracted from https://github.com/solana-labs/solana/pull/31986 to not include other small refactorings and new tests which will be PR-d separately.

### CODE REVIEW

File: programs/bpf_loader/src/syscalls/cpi.rs [hunk 1]
```diff
@@ -1223,6 +1268,63 @@ fn update_callee_account(
...
+) -> Result<(), Error> {
+    let CallerAccount {
+        original_data_len,
+        vm_data_addr,
+        ..
+    } = caller_account;
+
+    let data_region = account_data_region(memory_mapping, *vm_data_addr, *original_data_len)?;
+    if let Some(region) = data_region {
+        match (
```
Comment (Lichtso) [inline programs/bpf_loader/src/syscalls/cpi.rs]: Is the change / edge detection really useful here? Why not always set to the new state?

File: programs/bpf_loader/src/syscalls/cpi.rs [hunk 2]
```diff
@@ -1242,34 +1344,37 @@ fn update_caller_account(
...
-        let callee_ptr = callee_account.get_data().as_ptr() as u64;
-        if region.host_addr.get() != callee_ptr {
-            region.host_addr.set(callee_ptr);
+            // If an account's data pointer has changed - because of CoW or because
+            // of using AccountSharedData directly (deprecated) - we must update the
+            // corresponding MemoryRegion in the caller's address space. Address
+            // spaces are fixed so we don't need to update the MemoryRegion's
+            // length.
+            let callee_ptr = callee_account.get_data().as_ptr() as u64;
+            if region.host_addr.get() != callee_ptr {
```
Comment (Lichtso) [inline programs/bpf_loader/src/syscalls/cpi.rs:1369]: Again, why not always update the host address?

Comment (alessandrod) [reply programs/bpf_loader/src/syscalls/cpi.rs:1369]: see two lines below:
```
zero_all_spare_capacity = true
```

We need to detect when the pointer changes

File: programs/bpf_loader/src/syscalls/cpi.rs [hunk 3]
```diff
@@ -9,14 +13,62 @@ use {
...
+        debug!(
+            "Invalid account info pointer `{}': {:#x} != {:#x}",
+            field, vm_addr, expected_vm_addr
+        );
+        return Err(SyscallError::InvalidPointer.into());
+    }
+    Ok(())
+}
+
+enum VmValue<'a, 'b, T> {
```
[the remark this answers is not in the document]
Comment (alessandrod) [reply programs/bpf_loader/src/syscalls/cpi.rs:37]: The Translated variant will go away yes, but I'm not sure if VmValue should go away altogether. I kinda think we should go in the opposite direction and use it for `lamports` and `owner` too, since it makes it explicit that they're vm memory and not host memory.

[the remark this answers is not in the document]
Comment (Lichtso) [reply programs/bpf_loader/src/syscalls/cpi.rs:37]: Just write a comment that it should turned from `enum` to `struct` once the feature gate is activated.

File: programs/bpf_loader/src/syscalls/cpi.rs (hunk 1 above)
Comment (alessandrod) [reply programs/bpf_loader/src/syscalls/cpi.rs]: I think you commented on an old commit, I'm going to resolve

File: programs/bpf_loader/src/syscalls/cpi.rs [hunk 4]
```diff
@@ -78,21 +163,46 @@ impl<'a> CallerAccount<'a> {
...
-            )? as *mut u64;
-            let ref_to_len_in_vm = unsafe { &mut *translated };
+            let ref_to_len_in_vm = if direct_mapping {
+                let vm_addr = (account_info.data.as_ptr() as *const u64 as u64)
+                    .saturating_add(size_of::<u64>() as u64);
+                // In the same vein as the other check_account_info_pointer()
+                // checks, we don't lock this pointer to a specific address but we
+                // don't want it to be inside accounts, or calles might be able to
+                // write to the pointed memory.
+                if vm_addr >= ebpf::MM_INPUT_START {
```
Comment (ryoqun) [inline programs/bpf_loader/src/syscalls/cpi.rs:188]: seems quite robust guard...

File: programs/bpf_loader/src/syscalls/cpi.rs [hunk 5]
```diff
@@ -763,7 +930,7 @@ where
         } else if let Some(caller_account_index) =
             account_info_keys.iter().position(|key| *key == account_key)
         {
-            let original_data_len = accounts_metadata
+            let serialied_metadata = accounts_metadata
```
Comment (ryoqun) [inline programs/bpf_loader/src/syscalls/cpi.rs:933]: typo: `serialized_metadata` (same as this: https://github.com/solana-labs/solana/pull/31986/files#r1279270148)

File: programs/bpf_loader/src/syscalls/cpi.rs [hunk 6]
```diff
@@ -28,43 +88,68 @@ struct CallerAccount<'a> {
...
+            )?;
+            check_account_info_pointer(
+                invoke_context,
+                account_info.owner as *const _ as u64,
+                account_metadata.vm_owner_addr,
+                "owner",
+            )?;
+        }
+
+        let original_data_len = account_metadata.original_data_len;
```
Comment (ryoqun) [inline programs/bpf_loader/src/syscalls/cpi.rs:132]: nit: seems this variable is only used at the very last of just before returning from this fn. maybe, this var can be removed altogether?

File: programs/bpf_loader/src/syscalls/cpi.rs [hunk 7]
```diff
@@ -1184,14 +1450,57 @@ fn update_caller_account(
...
+
+                    // Temporarily configure the realloc region as writable then set it back to
+                    // whatever state it had.
+                    let realloc_region = caller_account
+                        .realloc_region(memory_mapping, is_loader_deprecated)?
+                        .unwrap(); // unwrapping here is fine, we already asserted !is_loader_deprecated
+                    let original_state = realloc_region.state.replace(MemoryState::Writable);
+                    defer!(
+                        realloc_region.state.set(original_state);
+                    );
```
Comment (ryoqun) [inline programs/bpf_loader/src/syscalls/cpi.rs:1486]: because of the existence of `;` (which is good by its own), i'd prefer `defer! {...}` like this:

```rust
defer! {
    realloc_region.state.set(original_state);
}
```

File: programs/bpf_loader/src/syscalls/cpi.rs [hunk 8]
```diff
@@ -1245,21 +1553,87 @@ fn update_caller_account(
...
+            //
+            // Therefore we temporarily configure the realloc region as writable
+            // then set it back to whatever state it had.
+            let realloc_region = caller_account
+                .realloc_region(memory_mapping, is_loader_deprecated)?
+                .unwrap(); // unwrapping here is fine, we asserted !is_loader_deprecated
+            let original_state = realloc_region.state.replace(MemoryState::Writable);
+            defer!(
+                realloc_region.state.set(original_state);
+            );
```
Comment (ryoqun) [inline programs/bpf_loader/src/syscalls/cpi.rs:1576]: same as https://github.com/solana-labs/solana/pull/32649/files?diff=split&w=0#r1307447593

**Comment by ryoqun** [comment]
status: phew, reviewing the dm is really hard... almost ready for lgtm with nits from me. really appreciate months of patience for code reviewing by me....

i need to cool down my brain a bit. i'll approve this and the other #31986 tomorrow unless tomorrow me beats tonight's me. :)

**Comment by alessandrod** [comment]
> status: phew, reviewing the dm is really hard... almost ready for lgtm with nits from me. really appreciate months of patience for code reviewing by me....
> 
> i need to cool down my brain a bit. i'll approve this and the other #31986 tomorrow unless tomorrow me beats tonight's me. :)

Thank you again for your review, I know this is complex code nobody wants to get near to 😅 I think I've addressed all  your nits, let me know if there's more

File: programs/bpf_loader/src/syscalls/cpi.rs [hunk 9]
```diff
@@ -991,6 +1153,25 @@ fn cpi_common<S: SyscallInvokeSigned>(
         .feature_set
         .is_active(&feature_set::bpf_account_data_direct_mapping::id());
 
+    // When direct mapping is on, we must update all perms at once before doing
+    // account data updates. Otherwise, updating an account might write into
+    // another account whose perms we haven't updated yet. See
+    // TEST_FORBID_LEN_UPDATE_AFTER_OWNERSHIP_CHANGE.
```
Comment (ryoqun) [inline programs/bpf_loader/src/syscalls/cpi.rs:1159]: nit: this comment is stale. it should be impossible by now, thanks to `check_account_info_pointer` and the new `>= ebpf::MM_INPUT_START` checks. when both guards are combined, there is no way to create aliasing...
that means, i think we can move this back into the original loop. however, i lean on leaving this as-is just to be safer side. so, how about updating comment at least?

also, maybe related to the ci failure: https://github.com/solana-labs/solana/pull/31986#issuecomment-1696792968

File: programs/bpf_loader/src/syscalls/cpi.rs [hunk 10]
```diff
@@ -1128,29 +1359,40 @@ fn update_caller_account(
...
-        // If an account's data pointer has changed - because of CoW or because
-        // of using AccountSharedData directly (deprecated) - we must update the
-        // corresponding MemoryRegion in the caller's address space. Address
-        // spaces are fixed so we don't need to update the MemoryRegion's length.
-        let region = memory_mapping.region(AccessType::Load, caller_account.vm_data_addr)?;
-        let callee_ptr = callee_account.get_data().as_ptr() as u64;
-        if region.host_addr.get() != callee_ptr {
-            region.host_addr.set(callee_ptr);
+            // If an account's data pointer has changed - because of CoW or because
+            // of using AccountSharedData directly (deprecated) - we must update the
```
Comment (ryoqun) [inline programs/bpf_loader/src/syscalls/cpi.rs:1381]: nit: how about mentioning the above `.reserve()` call as well for completeness?

```rust
    // If an account's data pointer has changed - because of CoW or because
    // of using AccountSharedData directly (deprecated) or just above `.reserve()` 
    // - we must update the
```

Comment (alessandrod) [reply programs/bpf_loader/src/syscalls/cpi.rs:1381]: updated

File: programs/bpf_loader/src/syscalls/cpi.rs [hunk 11]
```diff
@@ -1245,21 +1548,88 @@ fn update_caller_account(
...
+                invoke_context.get_check_size(),
+            )?
+        };
         let from_slice = callee_account
             .get_data()
             .get(caller_account.original_data_len..post_len)
             .ok_or(SyscallError::InvalidLength)?;
+        if to_slice.len() != from_slice.len() {
+            return Err(Box::new(InstructionError::AccountDataTooSmall));
+        }
```
Comment (ryoqun) [inline programs/bpf_loader/src/syscalls/cpi.rs:1589]: nit: i wonder how this error condition could be triggered, sans the integer math saturation (which can't happen in this context)....

to_slice.len() actually is `realloc_bytes_used`, which in turn is `post_len.saturating_sub(caller_account.original_data_len)`, and `from_slice.len()` is `caller_account.original_data_len..post_len`

maybe better off promoting to `assert!()` or demoting to `debug_assert!()`? (hint: I'm pro assert guy)

File: programs/bpf_loader/src/syscalls/cpi.rs [hunk 12]
```diff
@@ -1128,29 +1359,39 @@ fn update_caller_account(
...
+            // Since each instruction account is directly mapped in a memory region
+            // with a *fixed* length, upon returning from CPI we must ensure that the
+            // current capacity is at least the original length (what is mapped in
+            // memory), so that the account's memory region never points to an
+            // invalid address.
+            let min_capacity = caller_account.original_data_len;
+            if callee_account.capacity() < min_capacity {
+                callee_account.reserve(min_capacity.saturating_sub(callee_account.capacity()))?;
+                zero_all_spare_capacity = true;
+            }
```
Comment (ryoqun) [inline programs/bpf_loader/src/syscalls/cpi.rs:1378]: nit: hmm, this seems to overallocate capacity theoretically. the problem is that this additional capacity calc aren't considering the fact `.reserve(MAX_PERMITTED_DATA_INCREASE)` can theoretically called inside the invocation of `.make_data_mut()` quoted below:

```rust
    pub fn reserve(&mut self, additional: usize) -> Result<(), InstructionError> {
        // Note that we don't need to call can_data_be_changed() here nor
        // touch() the account. reserve() only changes the capacity of the
        // memory that holds the account but it doesn't actually change content
        // nor length of the account.
        self.make_data_mut();
        self.account.reserve(additional);
        Ok(())
    }
```

so MAX_PERMITTED_DATA_INCREASE bytes could be overallocated. but, in practice, this shouldn't be problem because `callee_account.capacity() < caller_account.original_data_len` imply `!is_shared()`.

that said, you can make this as-is , or make `.reserve()` adjust its `additional` after `.make_data_mut()` or even better adjust api like this (i know I'm a perfectionist...):

```rust
    pub fn ensure_capacity(&mut self, min_capacity: usize) -> bool {
        let mut current_capacity = self.capacity();
        if current_capacity >= min_capacity {
            return false;
        }
        if self.make_data_mut() {
            current_capacity = self.capacity();
        }

        let additional = min_capacity.saturating_sub(current_capacity);
        if additional == 0 {
            return false;
        }
        // Note that we don't need to call can_data_be_changed() here nor
        // touch() the account. reserve() only changes the capacity of the
        // memory that holds the account but it doesn't actually change content
        // nor length of the account.
        self.account.reserve(additional);
        true
    }
```

Comment (ryoqun) [reply programs/bpf_loader/src/syscalls/cpi.rs:1378]: ... and call site (the commented code here) would be like this:

```rust
zero_all_spare_capacity |= callee_account.ensure_capacity(caller_account.original_data_len);
```

Comment (alessandrod) [reply programs/bpf_loader/src/syscalls/cpi.rs:1378]: quoting from rust docs for Vec::reserve

> Reserves capacity for at least additional more elements to be inserted in the given Vec<T>. The collection may reserve more space to speculatively avoid frequent reallocations.

...which is exactly what we do? This seems better to me that adding more API, and for this particular case as you said the over allocation doesn't even happen

File: programs/bpf_loader/src/syscalls/cpi.rs [hunk 13]
```diff
@@ -1128,29 +1359,39 @@ fn update_caller_account(
...
-        // Since each instruction account is directly mapped in a memory region
-        // with a *fixed* length, upon returning from CPI we must ensure that the
-        // current capacity is at least the original length (what is mapped in
-        // memory), so that the account's memory region never points to an
-        // invalid address.
-        let min_capacity = caller_account.original_data_len;
-        if callee_account.capacity() < min_capacity {
-            callee_account.reserve(min_capacity.saturating_sub(callee_account.capacity()))?
-        }
+    let mut zero_all_spare_capacity = false;
```
Comment (ryoqun) [inline programs/bpf_loader/src/syscalls/cpi.rs:1362]: nit: actually not _all_; rather `should_zero_mapped_spare_capacity`.

Comment (ryoqun) [reply programs/bpf_loader/src/syscalls/cpi.rs:1362]: or just `should_zero_spare_capacity` (drop the overly qualifiying _all_); seems `spare` word already meaning `mapped_capacity`

File: programs/bpf_loader/src/syscalls/cpi.rs [hunk 14]
```diff
@@ -1128,29 +1359,39 @@ fn update_caller_account(
...
+                zero_all_spare_capacity = true;
+            }
         }
     }
-    let prev_len = *caller_account.ref_to_len_in_vm as usize;
+
+    let prev_len = *caller_account.ref_to_len_in_vm.get()? as usize;
     let post_len = callee_account.get_data().len();
+    let realloc_bytes_used = post_len.saturating_sub(caller_account.original_data_len);
     if prev_len != post_len {
```
Comment (ryoqun) [inline programs/bpf_loader/src/syscalls/cpi.rs:1395]: maybe, we should always zero spare capacity always here if `zero_all_spare_capacity == true`, regardless `post_len < prev_len`or not?

i mean, is it safe to skip zeroing conditionally while indicated to do so as such by the  flag?

say, account data vector could be changed by native program inside the called ix in the manner of Arc's Clone-on-Write fashion via `set_data_from_slice''s early-returning `.set_data()`, while still holding `prev_len == post_len`, right?

File: programs/bpf_loader/src/syscalls/cpi.rs (hunk 11 above)
Comment (alessandrod) [reply programs/bpf_loader/src/syscalls/cpi.rs:1589]: Unless $future_bugs, this condition can't happen, in the same way that the same check in the !direct_mapping path can't happen. It's a "just in case" defensive check, which as I mentioned I'm not in favor of having, but it doesn't seem like changing our policy belongs to this PR?

File: programs/bpf_loader/src/syscalls/cpi.rs [hunk 15]
```diff
@@ -1168,13 +1409,33 @@ fn update_caller_account(
...
+                    // difference between the previous and current lengths. The
+                    // rest of the memory contains whatever it contained before,
+                    // which is deterministic.
+                    prev_len
+                }
+                .saturating_sub(post_len);
+                if spare_len > 0 {
                     let dst = callee_account
                         .spare_data_capacity_mut()?
                         .get_mut(..spare_len)
```
Comment (ryoqun) [inline programs/bpf_loader/src/syscalls/cpi.rs:1441]: related to this https://github.com/solana-labs/solana/pull/32649/files#r1308918716

i think `zero_all_spare_capacity` should be used as an optimization hint, not part of critical code path; reasoning is very hard...

so, how about this?:

```rust
                if !zero_all_spare_capacity {
                    let spare_len = caller_account.original_data_len.saturating_sub(post_len);
                    let dst = callee_account
                        .spare_data_capacity_mut()?
                        .get_mut(..spare_len)
                        .ok_or_else(|| Box::new(InstructionError::AccountDataTooSmall))?
                        .as_mut_ptr();
                    // Safety: we check bounds above
                    unsafe { ptr::write_bytes(dst, 0, spare_len) };
                }
```

Comment (alessandrod) [reply programs/bpf_loader/src/syscalls/cpi.rs:1441]: The code above is for the zero_all_spare_capacity==true condition?

File: programs/bpf_loader/src/syscalls/cpi.rs (hunk 14 above)
Comment (ryoqun) [reply programs/bpf_loader/src/syscalls/cpi.rs:1395]: then, assume `{prev,post}_len < original_data_len`, on-chain program can see uninitialized memory in spare capacity?

hmm, wait, to create the precondition, CoW must have been performed already before the cpi?? so, actually this is impossible?

Comment (alessandrod) [reply programs/bpf_loader/src/syscalls/cpi.rs:1395]: > i mean, is it safe to skip zeroing conditionally while indicated to do so as such by the flag?

It is safe because `slice::to_vec()` is `Vec::with_capacity(slice.len())`. So if `slice.len() < original_data_len`, it'll hit the `if callee_account.capacity() < min_capacity {` case and set `zero_all_spare_capacity=true`.

I guess I can make this more explicit and inline `slice::to_vec()` in our code.

File: programs/bpf_loader/src/syscalls/cpi.rs [hunk 16]
```diff
@@ -1033,22 +1215,34 @@ fn update_callee_account(
...
+                    )?;
                     callee_account
                         .get_data_mut()?
                         .get_mut(caller_account.original_data_len..post_len)
                         .ok_or(SyscallError::InvalidLength)?
                         .copy_from_slice(
-                            caller_account
-                                .serialized_data
+                            serialized_data
                                 .get(0..realloc_bytes_used)
```
Comment (ryoqun) [inline programs/bpf_loader/src/syscalls/cpi.rs:1246]: https://github.com/solana-labs/solana/pull/31986#discussion_r1279363314

File: programs/bpf_loader/src/lib.rs [hunk 17]
```diff
@@ -1510,18 +1517,22 @@ fn execute<'a, 'b: 'a>(
...
+    // can map to a more specific error
+    let account_region_addrs = accounts_metadata
         .iter()
-        .map(|r| r.vm_addr..r.vm_addr.saturating_add(r.len))
+        .map(|m| {
+            let mut vm_end = m.vm_data_addr.saturating_add(m.original_data_len as u64);
+            if !is_loader_deprecated {
+                vm_end = vm_end.saturating_add(MAX_PERMITTED_DATA_INCREASE as u64)
+            }
+            m.vm_data_addr..vm_end
```
Comment (ryoqun) [inline programs/bpf_loader/src/lib.rs:1533]: https://github.com/solana-labs/solana/pull/31986#discussion_r1281301526

File: programs/bpf_loader/src/syscalls/cpi.rs (hunk 15 above)
Comment (ryoqun) [reply programs/bpf_loader/src/syscalls/cpi.rs:1441]: no. i was proposing to move the zeroing `if` clause for `zero_all_spare_capacity == true` out of here to the position at https://github.com/solana-labs/solana/pull/32649#discussion_r1308918716. The code is really for `zero_all_spare_capacity == false` to leave it here with some tweaks.

but then, the move seems not to be needed according to your explanation ( https://github.com/solana-labs/solana/pull/32649#discussion_r1308956943 ) and my correction ( https://github.com/solana-labs/solana/pull/32649#discussion_r1308942905 )

i think these statements are held true, assuming any given AccountSharedData never transition back to `is_shared == true` once after becoming `!is_shred`. in that case, i think conditional `zero_all_spare_capacity` zeroing (post_len < prev_len) is safe because there is no possibility of another CoW under the situation where `{prev,post}_len < original_data_len`... 

in other words, `{prev,post}_len < original_data_len` will be needed to expose uninitialized spare mem to on-chain programs with 2nd CoW with `!(post_len < prev_len)`. but this seems impossible to trigger in the first place...

**Review by ryoqun — APPROVED** [review]
lgtm with nits.

really thanks for the hard work.

**Review by Lichtso — APPROVED** [review]

**Comment by alessandrod** [comment]
> lgtm with nits.
> 
> really thanks for the hard work.

Thank you for the great review! 😊
